doc/specs/#4999 - Improved keyboard handling in Conpty.md

```diff
@@ -0,0 +1,452 @@
+
+## Solution Design
+
+### Inspiration
+
+The design we've settled upon is one that's highly inspired by a few precedents:
+* `Application Cursor Keys (DECCKM)` is a long-supported VT sequences which a
+  client application can use to request a different input format from the
+  Terminal. This is the DECSET sequence `^[[?1h`/`^[[?1l` (for enable/disable,
+  respectively). This changes the sequences sent by keys like the Arrow keys
+  from a sequence like `^[[A` to `^[OA` instead.
+* The `kitty` terminal emulator uses a similar DECSET sequence for enabling
+  their own input format, which they call ["full mode"]. Similar to DECCKM, this
+  changes the format of the sequences that the terminal should send for keyboard
+  input. Their "full mode" contains much more information when keys are pressed
+  or released (though, less than a full `INPUT_RECORD` worth of data). Instead
+  of input being sent to the client as a CSI or SS3 sequence, this `kitty` mode
+  uses "Application Program-Command" (or "APC) sequences , prefixed with `^[_`.
+* [iTerm2](https://www.iterm2.com/documentation-escape-codes.html) has a region
+  of OSC's that they've carved for themselves all starting with the same initial
+  parameter, `1337`. They then have a number of commands that all use the second
+  parameter to indicate what command specific to iTerm2 they're actually
+  implementing.
+
+### Requesting `win32-input-mode`
+
+An application can request `win32-input-mode` with the following OSC sequence
+
+```
+  ^[ ] 1000 ; 1 ; Ps ST
+
+         Ps: Whether to enable or disable win32-input-mode. If omitted, defaults to '0'.
+             0: Disable win32-input-mode
+             1: Enable win32-input-mode
+```
+
+OSC 1000 seemed to be unused according to [this discussion of used
+OSCs](https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/10).
+This seems like a reasonable place to stake a claim for sequences used by the
+the Windows Terminal - future WT-specific sequences could also be placed as `OSC
+1000;2`, `OSC 1000;3`, etc. This pattern is also similar to the way iTerm2
+defines their own sequences.
+
+When a terminal recieves a `^[]1000;1;1ST` sequence, they should switch into
+`win32-input-mode`. In `win32-input-mode`, the terminal will send keyboard input
+to the connected client application in the following format:
+
+### `win32-input-mode` sequences
+
+The `KEY_EVENT_RECORD` portion of an input record (the part that's important for
+us to encode in this feature) is defined as the following:
+
+```c++
+typedef struct _KEY_EVENT_RECORD {
+  BOOL  bKeyDown;
+  WORD  wRepeatCount;
+  WORD  wVirtualKeyCode;
+  WORD  wVirtualScanCode;
+  union {
+    WCHAR UnicodeChar;
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `!`. This character appears

```

Comment:
Note that `!` is not actually a valid final character for a CSI sequence - it's an intermediate. A standard VT parser would probably expect that to be followed by at least another character. For example, see [`DECSTR`](https://vt100.net/docs/vt510-rm/DECSTR.html).

Reply:
Oh, that's a good catch. Thanks! I totally forgot about those. I've update to `_`, which also looks unused.